types/lodash/fp/index.d.ts

```diff
@@ -0,0 +1,618 @@
+         * running each of its own enumerable properties through iteratee, with each invocation potentially mutating
+         * the accumulator object. The iteratee is bound to thisArg and invoked with four arguments: (accumulator,
+         * value, key, object). Iteratee functions may exit iteration early by explicitly returning false.
+         *
+         * @param iteratee The function invoked per iteration.
+         * @param object The object to iterate over.
+         * @param accumulator The custom accumulator value.
+         * @param thisArg The this binding of iteratee.
+         * @return Returns the accumulated value.
+         */
+        transform<T, TResult>(iteratee: T, accumulator: any): (object: T[]) => TResult[];
+
+        /**
+         * Checks if value is classified as an Array object.
+         * @param value The value to check.
+         *
+         * @return Returns true if value is correctly classified, else false.
+         */
+        isArray<T>(value?: any): value is T[];
+
+        /**
+         * Removes all elements from array that predicate returns truthy for and returns an array of the removed
+         * elements. The predicate is bound to thisArg and invoked with three arguments: (value, index, array).
+         *
+         * If a property name is provided for predicate the created _.property style callback returns the property
+         * value of the given element.
+         *
+         * If a value is also provided for thisArg the created _.matchesProperty style callback returns true for
+         * elements that have a matching property value, else false.
+         *
+         * If an object is provided for predicate the created _.matches style callback returns true for elements that
+         * have the properties of the given object, else false.
+         *
+         * Note: Unlike _.filter, this method mutates array.
+         *
+         * @param predicate The function invoked per iteration.
+         * @return Returns the new array of removed elements.
+         */
+        remove<T>(predicate: Function|Object): (array: List<T>) => T[];
+
+        /**
+         * Removes all provided values from array using SameValueZero for equality comparisons.
+         *
+         * Note: Unlike _.without, this method mutates array.
+         *
+         * @param array The array to modify.
+         * @param values The values to remove.
+         * @return Returns array.
+         */
+        pull<T>(array: T[], ...values: T[]): T[];
+
+        /**
+         * @see _.pull
+         */
+        pull<T>(array: List<T>, ...values: T[]): List<T>;
+
+        keys(arg: any);
+
+        range(min: number, max: number): number;
+
+        /**
+         * Recursively flattens a nested array.
+         *
+         * @param array The array to recursively flatten.
+         * @return Returns the new flattened array.
+         */
+        flattenDeep<T>(array: T): T[];
```

Comment:
Is this type correct?

Reply:
I believe that at the moment it is prohibitively difficult to successfully give the correct type here. The type would need to recur arbitrarily to determine the base type. Take this for example:
```ts
const x = flattenDeep([[1, 2], 3, 4])
```
`x` should be of type number, but here it would be `(number | number[])[][]`.

---

For a simple one level flatten, we could do:
```ts
flatten<T extends any[]>(array: T): T[number];
```
But this type fails as soon as we try to pass in a deep array.

Reply:
I understand the difficulty, I just meant that how it's currently typed is probably misleading:
`T => T[]`. If anything, `T[] => T` would make a bit more sense, but as you point out, you cannot really assume that type either.

Perhaps this would be better?
```
flattenDeep<T extends any[]>(nestedArray: any[]): T[];
```

Reply:
I definitely agree that its current form doesn't cut it. Unfortunately using the type param at the end just gives us a fancy form of type assertion, and won't actually buy us anything. In the end, I think the best we could do would be:
```ts
flattenDeep(array: any[]): any[];
```
because any other assumption we make here isn't going to work more generically.